src/internal/obj/local_client.go

```diff
@@ -1,126 +1,149 @@
-	"github.com/pachyderm/pachyderm/v2/src/internal/tracing"
+	"github.com/pachyderm/pachyderm/v2/src/internal/uuid"
+	"github.com/sirupsen/logrus"
 )
 
 // NewLocalClient returns a Client that stores data on the local file system
-func NewLocalClient(root string) (Client, error) {
-	if err := os.MkdirAll(root, 0755); err != nil {
-		return nil, errors.EnsureStack(err)
-	}
+func NewLocalClient(rootDir string) (Client, error) {
 	var c Client
-	c = &localClient{filepath.Clean(root)}
+	var err error
+	c, err = newFSClient(rootDir)
+	if err != nil {
+		return nil, err
+	}
 	if monkeyTest {
 		c = &monkeyClient{c}
 	}
 	return newUniformClient(c), nil
 }
 
-type localClient struct {
-	root string
+type fsClient struct {
+	rootDir string
 }
 
-func (c *localClient) normPath(path string) string {
-	path = filepath.Clean(path)
-	return filepath.Join(c.root, path)
+func newFSClient(rootDir string) (Client, error) {
+	c := &fsClient{rootDir: filepath.Clean(rootDir)}
+	if c.rootDir == "" || c.rootDir == "/" || c.rootDir == "." {
+		panic("you probably didn't want to set the local client's root path to " + c.rootDir)
+	}
+	return c, c.init()
 }
 
-func (c *localClient) Put(_ context.Context, path string, r io.Reader) (retErr error) {
-	defer func() { retErr = c.transformError(retErr, path) }()
-	fullPath := c.normPath(path)
-
-	// Create the directory since it may not exist
-	if err := os.MkdirAll(filepath.Dir(fullPath), 0755); err != nil {
+func (c *fsClient) Put(ctx context.Context, name string, r io.Reader) error {
+	staging := c.stagingPathFor(name)
+	final := c.finalPathFor(name)
+	f, err := os.Create(staging)
+	if err != nil {
 		return err
 	}
-	file, err := os.Create(fullPath)
-	if err != nil {
+	defer f.Close()
+	if _, err := io.Copy(f, r); err != nil {
 		return err
 	}
-	defer func() {
-		if err := file.Close(); retErr == nil {
-			retErr = err
-		}
-	}()
-	if _, err := io.Copy(file, r); err != nil {
+	if err := f.Close(); err != nil {
 		return err
 	}
-	return nil
+	return os.Rename(staging, final)
 }
 
-func (c *localClient) Get(_ context.Context, path string, w io.Writer) (retErr error) {
-	defer func() { retErr = c.transformError(retErr, path) }()
-	file, err := os.Open(c.normPath(path))
+func (c *fsClient) Get(ctx context.Context, name string, w io.Writer) (retErr error) {
+	defer func() { retErr = c.transformError(retErr, name) }()
+	f, err := os.Open(c.finalPathFor(name))
 	if err != nil {
 		return err
 	}
-	defer func() {
-		if err := file.Close(); retErr == nil {
-			retErr = err
-		}
-	}()
-	_, err = io.Copy(w, file)
+	defer f.Close()
+	_, err = io.Copy(w, f)
 	return err
 }
 
-func (c *localClient) Delete(_ context.Context, path string) (retErr error) {
-	defer func() { retErr = c.transformError(retErr, path) }()
-	return os.Remove(c.normPath(path))
+func (c *fsClient) Delete(ctx context.Context, name string) error {
+	err := os.Remove(c.finalPathFor(name))
+	if os.IsNotExist(err) {
```

Comment:
Is this normal object storage semantics?

Reply:
The behavior varies in terms of whether Delete is idempotent.  We enforce this behavior again in the uniform client.